File: py/lib/crypto.py
"""Cryptographic helpers."""

import binascii
import json
import re
import secrets
import struct
from typing import Any, IO, List, Optional, Tuple, Union

# pip install base58
import base58

from Cryptodome.Cipher import AES
from Cryptodome.Hash import RIPEMD160, SHA3_256, SHA3_512

# pylint: disable=unused-import
from ecdsa.keys import BadSignatureError
# pylint: enable=unused-import

# pip install ecdsa
from ecdsa import SigningKey, VerifyingKey, SECP256k1


def my_random_bytes(num_bytes: int) -> bytes:
    """Return a random byte string."""
    return secrets.token_bytes(num_bytes)


class KeyValidationError(Exception):
    """Key Verification Failed."""

    pass


class Hash:
    """
    Abstraction for working with hashes.

    It turns out that hashes are a major source of confusion in that they can
    be represented in different ways, and that representation is vitally
    important. For example, do you want a bytes object where each byte
    contains just the binary data (with a length of 64 bytes) or do
    you want a string containing the printable representation of those
    characters, with a length of 128 bytes? And do you want that string
    representation to be an ascii byte string or a unicode string?

    Getting this right is paramount to a working cryptocurrency, and so this
    object is intended to enforce the correct representation where possible.
    """

    def __init__(self, hash_input: Union[str, bytes], bits=512) -> None:
        """
        Create a new Hash object.

        The hash will be auto-inferred based on the type and length of the
        input. The 'bits' parameter is used to determine how many bytes of
        data the hash actually contains. The default is 512, corresponding to
        the SHA3_512 hashes used throughout this project.

        Internally the hash will be stored as bytes containing the binary data,
        since that is what is used in most hash operations.
        """
        num_bytes = int(bits/8)
        self.bits = bits
        self.data = b"0" * num_bytes
        hashlen = len(hash_input)

        # A unicode string is always understood to be a printable
        # representation of a hash. It should only contain ascii characters.
        if isinstance(hash_input, str):
            assert hashlen == (num_bytes * 2), \
                "Unicode hash string length is incorrect. " \
                "Got {}, expected {}".format(hashlen, num_bytes)
            self.data = binascii.unhexlify(hash_input.encode('ascii'))
        elif isinstance(hash_input, bytes):
            if hashlen == (num_bytes * 2):
                self.data = binascii.unhexlify(hash_input)
            elif hashlen == num_bytes:
                self.data = hash_input
            else:
                raise ValueError("Invalid hash length for '{}'. "
                                 "Got {}, expected {}".format(
                                     hash_input, hashlen, num_bytes))
        else:
            raise ValueError("Invalid hash format: {}".format(hash_input))
        return

    def __eq__(self, other: Any) -> bool:
        """Return True if other refers to the same hash."""
        if isinstance(other, Hash):
            return self.data == other.data

        # If other is not a hash, this will assert.
        return self.data == Hash(other, self.bits).data

    def __hash__(self):
        """Get the object hash to allow for comparison in sets."""
        return hash(self.data)

    def serialise(self) -> str:
        """
        Serialise to a unicode string.

        Since hashes are sometimes serialised to JSON, we use unicode strings
        to serialise them.
        """
        return self.to_string()

    def to_string(self) -> str:
        """Get the printable representation of the hash as a unicode string."""
        return binascii.hexlify(self.data).decode('ascii')

    def digest(self) -> bytes:
        """Return the raw binary digest. Use this for all hash operations."""
        return self.data


class Hash512(Hash):
    """Wrapper for a 512-bit hash, as used by Hasher/SHA3_512."""

    def __init__(self, hash_input: Union[str, bytes]) -> None:
        """Create a new Hash512 object."""
        if not hash_input:
            # Convert blank hash to SHA3.
            hash_input = Hasher(b"").get_hash().to_string()

        super().__init__(hash_input, bits=512)
        return

    @staticmethod
    def deserialise(data: str) -> 'Hash512':
        """Deserialise the string into a Hash512 object."""
        return Hash512(data)


class Hash256(Hash):
    """Wrapper for a 256-bit hash, as used by ECDSA keys."""

    def __init__(self, hash_input: Union[str, bytes]) -> None:
        """Create a new Hash256 object."""
        super().__init__(hash_input, bits=256)
        return

    @staticmethod
    def deserialise(data: str) -> 'Hash256':
        """Deserialise the string into a Hash256 object."""
        return Hash256(data)


def basic_hash_check(h: Union[str, Hash]) -> None:
    """Assert that the hash consists of alphanumeric characters only."""
    if isinstance(h, Hash):
        h = h.to_string()
    assert re.search(r'^[a-fA-F0-9]+$', h), "Hash check failed"
    return


class CryptoAddress:
    """
    A specific type used to store the base58 wallet addresses.

    The primary purpose of this class is to differentiate addresses from
    regular hashes, including facilitating type safety checks.
    """

    def __init__(self, address: str) -> None:
        """Create a new CryptoAddress object."""
        self.address = str(address)
        return

    def __eq__(self, other: Any) -> bool:
        """Check for equality."""
        if isinstance(other, CryptoAddress):
            return self.address == other.address

        return self.address == other

    def __hash__(self):
        """Get the object hash to allow for comparison in sets."""
        return hash(self.address)

    def serialise(self) -> str:
        """Serialise the address."""
        return self.address  # It's already a string.

    @staticmethod
    def deserialise(data: str) -> 'CryptoAddress':
        """Deserialise the string into a CryptoAddress object."""
        return CryptoAddress(data)

    def to_string(self) -> str:
        """Return string representation."""
        return self.address


class EncryptedData:
    """Helper for encrypted data."""

    def __init__(self, nonce: bytes, tag: bytes,
                 encrypted_data: bytes) -> None:
        """Create a new EncryptedData object."""
        self.nonce = nonce
        self.tag = tag
        self.encrypted_data = encrypted_data
        return

    def decrypt(self, key: Hash256) -> bytes:
        """Decrypt the data using the specified key."""
        cipher = AES.new(key.digest(), AES.MODE_EAX, self.nonce)
        ciphertext = cipher.decrypt_and_verify(self.encrypted_data, self.tag)
        return ciphertext

    def decrypt_with_passphrase(self, passphrase: str) -> bytes:
        """Decrypt the data using the specified passphrase."""
        key = Hash256(SHA3_256.new(data=passphrase.encode('utf-8')).digest())
        return self.decrypt(key)

    def decrypt_json(self, key: Hash256) -> Any:
        """Decrypt and unpack JSON, returning data structure."""
        json_str = self.decrypt(key)
        return json.loads(json_str)

    def write_to_file(self, filename: str,
                      overwrite_if_exists: bool = False) -> None:
        """Write the encrypted data to a file."""
        mode = "xb"
        if overwrite_if_exists:
            mode = "wb"

        with open(filename, mode) as f:
            self._write_bytes(f, self.nonce)
            self._write_bytes(f, self.tag)
            self._write_bytes(f, self.encrypted_data)
        return

    def _write_bytes(self, f: IO, b: bytes) -> None:
        """Write bytes to a file as length + data."""
        f.write(struct.pack("<Q", len(b)))
        f.write(b)
        return

    @staticmethod
    def read_from_file(filename: str) -> 'EncryptedData':
        """Read encrypted data from a file and return EncryptedData."""
        # Silence pydocstyle D202.

        def _read_bytes(f: IO) -> bytes:
            """Read a byte string from the current file position."""
            size = struct.unpack("<Q", f.read(8))[0]
            return f.read(size)

        with open(filename, "rb") as f:
            nonce = _read_bytes(f)
            tag = _read_bytes(f)
            data = _read_bytes(f)
            return EncryptedData(nonce=nonce, tag=tag, encrypted_data=data)


class CryptHelper:
    """Crypto helper class."""

    def __init__(self):
        """Create a new CryptHelper object (N/A)."""
        assert False, "Use static methods, not new()"
        return

    @staticmethod
    def encrypt(data: bytes) -> Tuple[Hash256, EncryptedData]:
        """Encrypt some data with a randomly generated key."""
        key = Hash256(my_random_bytes(32))
        cipher = AES.new(key.digest(), AES.MODE_EAX)
        ciphertext, tag = cipher.encrypt_and_digest(data)
        return key, EncryptedData(cipher.nonce, tag, ciphertext)

    @staticmethod
    def encrypt_json(data: Any) -> Tuple[Hash256, EncryptedData]:
        """Encrypt some object after first converting to JSON."""
        return CryptHelper.encrypt(json.dumps(data).encode('utf-8'))

    @staticmethod
    def encrypt_with_passphrase(data: bytes, passphrase: str) -> EncryptedData:
        """Encrypt some data with the specified password."""
        key = SHA3_256.new(data=passphrase.encode('utf-8')).digest()
        cipher = AES.new(key, AES.MODE_EAX)
        ciphertext, tag = cipher.encrypt_and_digest(data)
        return EncryptedData(cipher.nonce, tag, ciphertext)

    @staticmethod
    def get_hash_of_list(data_list: List[Union[bytes, str]]) -> Hash512:
        """Get the crypto hash for the specified list of items."""
        h = SHA3_512.new()
        for item in data_list:
            if isinstance(item, bytes):
                h.update(item)
            elif isinstance(item, str):
                h.update(item.encode('utf-8'))
            else:
                assert False, \
                    "Item must be bytes or str, but got {!r}".format(item)
        return Hash512(h.digest())


class ECDSAPublicKey:
    """Wrapper around a public ECDSA (verifying) key."""

    def __init__(self, pk: Union[VerifyingKey, bytes, str]) -> None:
        """Create a new ECDSAPublicKey object."""
        assert pk, "No pk specified!"

        if isinstance(pk, Hash512):
            self.key = VerifyingKey.from_string(pk.digest(),
                                                curve=SECP256k1)
        elif isinstance(pk, Hash):
            raise TypeError("Invalid hash type: {}".format(pk.to_string()))
        elif isinstance(pk, (str, bytes)):
            self.key = VerifyingKey.from_string(pk, curve=SECP256k1)
        else:
            self.key = pk
        return

    def as_hash(self) -> Hash512:
        """Get the key as a Hash512 object."""
        return Hash512(self.key.to_string())

    def as_pem(self) -> bytes:
        """Get the public key as a PEM format byte string."""
        return self.key.to_pem()

    def get_address(self) -> CryptoAddress:
        """Get the address corresponding to this public key, as bytes."""
        # NOTE: Use SHA3_256 to produce a different address than bitcoin.
        # If SHA256 is used, a bitcoin address would be produced.
        h = SHA3_256.new(self.key.to_string()).digest()
        r = RIPEMD160.new(h).digest()
        rv = b'\0' + r
        return CryptoAddress(base58.b58encode_check(rv))


class ECDSAPrivateKey:
    """Wrapper around a private ECDSA (signing) key."""

    def __init__(self, pk: Union[SigningKey, bytes, str]) -> None:
        """Create a new ECDSAPublicKey object."""
        assert pk, "No pk specified!"

        if isinstance(pk, (str, bytes)):
            self.key = SigningKey.from_string(pk, curve=SECP256k1)
        else:
            self.key = pk
        return

    @staticmethod
    def generate() -> 'ECDSAPrivateKey':
        """Generate a new key pair."""
        return ECDSAPrivateKey(pk=SigningKey.generate(curve=SECP256k1))

    @staticmethod
    def from_file(filename: str, passphrase: str) -> 'ECDSAPrivateKey':
        """Read from file (as written by write_key_pair)."""
        data = EncryptedData.read_from_file(filename)
        key = binascii.unhexlify(data.decrypt_with_passphrase(passphrase))
        return ECDSAPrivateKey(SigningKey.from_string(key, curve=SECP256k1))

    @property
    def publickey(self) -> 'ECDSAPublicKey':
        """Get the public key."""
        return ECDSAPublicKey(pk=self.key.get_verifying_key())

    @staticmethod
    def from_string(s) -> 'ECDSAPrivateKey':
        """Convert string to ECDSAPrivateKey object."""
        return ECDSAPrivateKey(binascii.unhexlify(s))

    def to_string(self):
        """
        Get the key as a string.

        WARNING: Do not store this to disk or send it over a network
        unencrypted.
        """
        return binascii.hexlify(self.key.to_string()).decode('utf-8')

    def write_key_pair(self, filename_prefix: str, passphrase: str) \
            -> Tuple[str, str]:
        """Write key pair to the specified filename prefix."""
        assert len(passphrase) >= 8, "Passphrase must be at least 8 chars!"

        fname_private = filename_prefix + "_private.pem"
        fname_public = filename_prefix + "_public.pem"

        key = binascii.hexlify(self.key.to_string())
        data = CryptHelper.encrypt_with_passphrase(
            data=key, passphrase=passphrase)
        data.write_to_file(fname_private)

        with open(fname_public, 'xt') as fpub:
            fpub.write(self.publickey.as_pem().decode('utf-8'))
        return fname_private, fname_public


# Type def for Hasher.update().
Hashable = Union[bytes, str, Hash]
HashableList = Union[Hashable, List[Hashable]]


class Hasher:
    """Hasher object to work with hashes."""

    def __init__(self, message_or_list: Optional[HashableList] = None) -> None:
        """Create a new Hasher object."""
        self.h = SHA3_512.new()
        if message_or_list:
            self.update(message_or_list)
        return

    def update(self, message_or_list: HashableList) -> 'Hasher':
        """Add message to hash."""
        if not isinstance(message_or_list, list):
            message_or_list = [message_or_list]  # type: ignore

        for msg in message_or_list:
            if isinstance(msg, Hash):
                self.h.update(msg.digest())
            elif isinstance(msg, bytes):
                self.h.update(msg)
            elif isinstance(msg, str):
                self.h.update(msg.encode('utf-8'))
            else:
                assert False, \
                    "Item must be bytes or str, but got {!r}".format(msg)
        return self  # Allow chaining.

    def get_hash(self) -> Hash512:
        """Get the resulting hash as a Hash512 object."""
        return Hash512(self.h.digest())

    def base58(self) -> bytes:
        """Get the base58 encoded bytes for this hash."""
        return base58.b58encode(self.h.digest())

    def sign(self, private_key: ECDSAPrivateKey) -> Hash512:
        """Sign the hash with the private key."""
        assert isinstance(private_key, ECDSAPrivateKey), \
            "Signing a hash requires a private key of type ECDSAPrivateKey"

        sig_raw = private_key.key.sign(self.h.digest())
        return Hash512(sig_raw)

    def verify_signature(self, sig: Hash512, public_key: ECDSAPublicKey) \
            -> None:
        """
        Verify that the signature was signed with the private key.

        This also verifies that the signature matches the hashed data.
        """
        try:
            assert isinstance(public_key, ECDSAPublicKey), \
                "Verifying a hash requires a public key of type ECDSAPublicKey"

            assert isinstance(sig, Hash512), "sig must be a Hash256 object"

            if not bool(public_key.key.verify(sig.digest(), self.h.digest())):
                raise KeyValidationError("Signature Verification Failed")
            return
        except AssertionError as exc:
            raise KeyValidationError("Invalid signature") from exc

File: rs/src/crypto.rs
/// Cryptographic helpers — direct port of lib/crypto.py.
///
/// Provides Hash types, Hasher (SHA3-512), ECDSA key pairs (secp256k1),
/// AES-EAX encryption, and CryptoAddress (base58check).
use std::fmt;
use std::fs::File;
use std::io::{Read, Write};

use aes::Aes256;
use digest::Digest;
use eax::aead::{AeadInPlace, KeyInit};
use eax::Eax;
use k256::ecdsa::signature::Verifier;
use k256::ecdsa::{Signature, SigningKey, VerifyingKey};
use rand::RngCore;
use ripemd::Ripemd160;
use sha3::{Sha3_256, Sha3_512};

use crate::errors::{BlockChainError, Result};

// ---------------------------------------------------------------------------
// Random bytes
// ---------------------------------------------------------------------------

pub fn random_bytes(n: usize) -> Vec<u8> {
    let mut buf = vec![0u8; n];
    rand::thread_rng().fill_bytes(&mut buf);
    buf
}

// ---------------------------------------------------------------------------
// Hash types
// ---------------------------------------------------------------------------

/// Generic hash wrapper storing raw bytes.
/// Python equivalent: Hash / Hash512 / Hash256
#[derive(Clone, Eq)]
pub struct Hash {
    pub data: Vec<u8>,
    pub bits: usize,
}

impl Hash {
    pub fn new(input: &[u8], bits: usize) -> Result<Self> {
        let num_bytes = bits / 8;
        let data = if input.len() == num_bytes * 2 {
            // Hex-encoded
            hex::decode(input).map_err(|e| {
                BlockChainError::Crypto(format!("Invalid hex hash: {e}"))
            })?
        } else if input.len() == num_bytes {
            // Raw bytes
            input.to_vec()
        } else {
            return Err(BlockChainError::Crypto(format!(
                "Invalid hash length: got {}, expected {} or {}",
                input.len(),
                num_bytes,
                num_bytes * 2,
            )));
        };
        Ok(Hash { data, bits })
    }

    pub fn from_hex(hex_str: &str, bits: usize) -> Result<Self> {
        Self::new(hex_str.as_bytes(), bits)
    }

    pub fn serialise(&self) -> String {
        self.to_hex()
    }

    pub fn to_hex(&self) -> String {
        hex::encode(&self.data)
    }

    pub fn digest(&self) -> &[u8] {
        &self.data
    }
}

impl PartialEq for Hash {
    fn eq(&self, other: &Self) -> bool {
        self.data == other.data
    }
}

impl std::hash::Hash for Hash {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.data.hash(state);
    }
}

impl fmt::Debug for Hash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Hash({})", self.to_hex())
    }
}

// ---------------------------------------------------------------------------
// Hash512 / Hash256 convenience constructors
// ---------------------------------------------------------------------------

/// 512-bit hash (SHA3-512 output size).
pub fn hash512(input: &[u8]) -> Result<Hash> {
    if input.is_empty() {
        // Match Python: blank input -> SHA3-512 of empty bytes
        let h = Hasher::new_with_message(b"");
        return Ok(h.get_hash());
    }
    Hash::new(input, 512)
}

pub fn hash512_from_hex(hex_str: &str) -> Result<Hash> {
    if hex_str.is_empty() {
        let h = Hasher::new_with_message(b"");
        return Ok(h.get_hash());
    }
    Hash::from_hex(hex_str, 512)
}

pub fn hash256(input: &[u8]) -> Result<Hash> {
    Hash::new(input, 256)
}

// ---------------------------------------------------------------------------
// basic_hash_check
// ---------------------------------------------------------------------------

pub fn basic_hash_check(h: &Hash) -> Result<()> {
    let s = h.to_hex();
    if s.chars().all(|c| c.is_ascii_hexdigit()) {
        Ok(())
    } else {
        Err(BlockChainError::Crypto("Hash check failed".into()))
    }
}

// ---------------------------------------------------------------------------
// CryptoAddress — base58check wallet address
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Eq)]
pub struct CryptoAddress {
    pub address: String,
}

impl CryptoAddress {
    pub fn new(address: &str) -> Self {
        CryptoAddress {
            address: address.to_string(),
        }
    }

    pub fn serialise(&self) -> String {
        self.address.clone()
    }

    pub fn deserialise(data: &str) -> Self {
        CryptoAddress::new(data)
    }

    pub fn to_string_repr(&self) -> &str {
        &self.address
    }

    /// Validate that this is a proper base58check address.
    pub fn validate(&self) -> Result<()> {
        bs58::decode(&self.address)
            .with_check(None)
            .into_vec()
            .map_err(|e| {
                BlockChainError::TransactionValidation(format!(
                    "Invalid output address: {e}"
                ))
            })?;
        Ok(())
    }
}

impl PartialEq for CryptoAddress {
    fn eq(&self, other: &Self) -> bool {
        self.address == other.address
    }
}

impl std::hash::Hash for CryptoAddress {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.address.hash(state);
    }
}

impl PartialEq<str> for CryptoAddress {
    fn eq(&self, other: &str) -> bool {
        self.address == other
    }
}

// ---------------------------------------------------------------------------
// EncryptedData — AES-256-EAX encryption
// ---------------------------------------------------------------------------

pub struct EncryptedData {
    pub nonce: Vec<u8>,
    pub tag: Vec<u8>,
    pub encrypted_data: Vec<u8>,
}

impl EncryptedData {
    pub fn decrypt(&self, key: &Hash) -> Result<Vec<u8>> {
        let cipher = Eax::<Aes256>::new_from_slice(key.digest())
            .map_err(|e| BlockChainError::Crypto(format!("AES init failed: {e}")))?;

        let nonce = eax::aead::Nonce::<Eax<Aes256>>::from_slice(&self.nonce);
        let tag = eax::aead::Tag::<Eax<Aes256>>::from_slice(&self.tag);

        let mut buffer = self.encrypted_data.clone();
        cipher
            .decrypt_in_place_detached(nonce, b"", &mut buffer, tag)
            .map_err(|e| BlockChainError::Crypto(format!("Decryption failed: {e}")))?;
        Ok(buffer)
    }

    pub fn decrypt_with_passphrase(&self, passphrase: &str) -> Result<Vec<u8>> {
        let key_bytes = Sha3_256::digest(passphrase.as_bytes());
        let key = hash256(&key_bytes)?;
        self.decrypt(&key)
    }

    pub fn decrypt_json(&self, key: &Hash) -> Result<serde_json::Value> {
        let plaintext = self.decrypt(key)?;
        let v = serde_json::from_slice(&plaintext)?;
        Ok(v)
    }

    pub fn write_to_file(&self, filename: &str, overwrite: bool) -> Result<()> {
        let mut f = if overwrite {
            File::create(filename)?
        } else {
            File::options()
                .write(true)
                .create_new(true)
                .open(filename)?
        };
        write_length_prefixed(&mut f, &self.nonce)?;
        write_length_prefixed(&mut f, &self.tag)?;
        write_length_prefixed(&mut f, &self.encrypted_data)?;
        Ok(())
    }

    pub fn read_from_file(filename: &str) -> Result<Self> {
        let mut f = File::open(filename)?;
        let nonce = read_length_prefixed(&mut f)?;
        let tag = read_length_prefixed(&mut f)?;
        let encrypted_data = read_length_prefixed(&mut f)?;
        Ok(EncryptedData {
            nonce,
            tag,
            encrypted_data,
        })
    }
}

fn write_length_prefixed(f: &mut File, data: &[u8]) -> Result<()> {
    let len = (data.len() as u64).to_le_bytes();
    f.write_all(&len)?;
    f.write_all(data)?;
    Ok(())
}

fn read_length_prefixed(f: &mut File) -> Result<Vec<u8>> {
    let mut len_buf = [0u8; 8];
    f.read_exact(&mut len_buf)?;
    let len = u64::from_le_bytes(len_buf) as usize;
    let mut buf = vec![0u8; len];
    f.read_exact(&mut buf)?;
    Ok(buf)
}

// ---------------------------------------------------------------------------
// CryptHelper — static encryption helpers
// ---------------------------------------------------------------------------

pub struct CryptHelper;

impl CryptHelper {
    pub fn encrypt(data: &[u8]) -> Result<(Hash, EncryptedData)> {
        let key = hash256(&random_bytes(32))?;
        let cipher = Eax::<Aes256>::new_from_slice(key.digest())
            .map_err(|e| BlockChainError::Crypto(format!("AES init failed: {e}")))?;

        let nonce_bytes = random_bytes(16);
        let nonce = eax::aead::Nonce::<Eax<Aes256>>::from_slice(&nonce_bytes);

        let mut buffer = data.to_vec();
        let tag = cipher
            .encrypt_in_place_detached(nonce, b"", &mut buffer)
            .map_err(|e| BlockChainError::Crypto(format!("Encryption failed: {e}")))?;

        Ok((
            key,
            EncryptedData {
                nonce: nonce_bytes,
                tag: tag.to_vec(),
                encrypted_data: buffer,
            },
        ))
    }

    pub fn encrypt_json(data: &serde_json::Value) -> Result<(Hash, EncryptedData)> {
        let json_bytes = serde_json::to_vec(data)?;
        Self::encrypt(&json_bytes)
    }

    pub fn encrypt_with_passphrase(data: &[u8], passphrase: &str) -> Result<EncryptedData> {
        let key_bytes = Sha3_256::digest(passphrase.as_bytes());
        let cipher = Eax::<Aes256>::new_from_slice(&key_bytes)
            .map_err(|e| BlockChainError::Crypto(format!("AES init failed: {e}")))?;

        let nonce_bytes = random_bytes(16);
        let nonce = eax::aead::Nonce::<Eax<Aes256>>::from_slice(&nonce_bytes);

        let mut buffer = data.to_vec();
        let tag = cipher
            .encrypt_in_place_detached(nonce, b"", &mut buffer)
            .map_err(|e| BlockChainError::Crypto(format!("Encryption failed: {e}")))?;

        Ok(EncryptedData {
            nonce: nonce_bytes,
            tag: tag.to_vec(),
            encrypted_data: buffer,
        })
    }
}

// ---------------------------------------------------------------------------
// ECDSAPublicKey — secp256k1 verifying key
// ---------------------------------------------------------------------------

#[derive(Clone, Debug)]
pub struct ECDSAPublicKey {
    key: VerifyingKey,
}

impl ECDSAPublicKey {
    /// Create from a Hash512 containing the raw 64-byte public key (x||y).
    pub fn from_hash(h: &Hash) -> Result<Self> {
        let raw = h.digest();
        if raw.len() != 64 {
            return Err(BlockChainError::KeyValidation(format!(
                "Expected 64-byte public key, got {}",
                raw.len()
            )));
        }
        // Prepend 0x04 for uncompressed SEC1 format
        let mut uncompressed = vec![0x04];
        uncompressed.extend_from_slice(raw);
        let key = VerifyingKey::from_sec1_bytes(&uncompressed)
            .map_err(|e| BlockChainError::KeyValidation(format!("Invalid public key: {e}")))?;
        Ok(ECDSAPublicKey { key })
    }

    pub fn from_verifying_key(key: VerifyingKey) -> Self {
        ECDSAPublicKey { key }
    }

    /// Get the raw 64-byte public key as a Hash512.
    pub fn as_hash(&self) -> Hash {
        let point = self.key.to_encoded_point(false);
        let bytes = point.as_bytes();
        // Skip 0x04 prefix to get 64-byte x||y
        Hash {
            data: bytes[1..].to_vec(),
            bits: 512,
        }
    }

    /// Get PEM-encoded public key.
    pub fn as_pem(&self) -> String {
        // Simplified: just hex-encode for the port
        // The Python version uses ECDSA PEM format
        let point = self.key.to_encoded_point(false);
        hex::encode(point.as_bytes())
    }

    /// Derive the base58check wallet address.
    /// Uses SHA3-256 (not SHA-256, intentionally different from Bitcoin).
    pub fn get_address(&self) -> CryptoAddress {
        let raw_key = self.key.to_encoded_point(false);
        // Hash with SHA3-256 (not SHA-256)
        let sha3_hash = Sha3_256::digest(&raw_key.as_bytes()[1..]);
        // Then RIPEMD-160
        let ripe_hash = Ripemd160::digest(&sha3_hash);
        // Prepend version byte 0x00
        let mut payload = vec![0x00];
        payload.extend_from_slice(&ripe_hash);
        // base58check encode
        let address = bs58::encode(&payload).with_check().into_string();
        CryptoAddress::new(&address)
    }

    pub fn verifying_key(&self) -> &VerifyingKey {
        &self.key
    }
}

// ---------------------------------------------------------------------------
// ECDSAPrivateKey — secp256k1 signing key
// ---------------------------------------------------------------------------

pub struct ECDSAPrivateKey {
    key: SigningKey,
}

impl ECDSAPrivateKey {
    pub fn generate() -> Self {
        ECDSAPrivateKey {
            key: SigningKey::random(&mut rand::thread_rng()),
        }
    }

    pub fn from_bytes(bytes: &[u8]) -> Result<Self> {
        let key = SigningKey::from_slice(bytes)
            .map_err(|e| BlockChainError::KeyValidation(format!("Invalid private key: {e}")))?;
        Ok(ECDSAPrivateKey { key })
    }

    pub fn from_string(hex_str: &str) -> Result<Self> {
        let bytes = hex::decode(hex_str)
            .map_err(|e| BlockChainError::KeyValidation(format!("Invalid hex: {e}")))?;
        Self::from_bytes(&bytes)
    }

    pub fn from_file(filename: &str, passphrase: &str) -> Result<Self> {
        let encrypted = EncryptedData::read_from_file(filename)?;
        let key_hex = encrypted.decrypt_with_passphrase(passphrase)?;
        let key_bytes = hex::decode(&key_hex)
            .map_err(|e| BlockChainError::KeyValidation(format!("Invalid key data: {e}")))?;
        let key = SigningKey::from_slice(&key_bytes)
            .map_err(|e| BlockChainError::KeyValidation(format!("Invalid private key: {e}")))?;
        Ok(ECDSAPrivateKey { key })
    }

    pub fn publickey(&self) -> ECDSAPublicKey {
        ECDSAPublicKey::from_verifying_key(*self.key.verifying_key())
    }

    pub fn to_hex_string(&self) -> String {
        hex::encode(self.key.to_bytes())
    }

    /// Write key pair to files. Returns (private_filename, public_filename).
    pub fn write_key_pair(
        &self,
        filename_prefix: &str,
        passphrase: &str,
    ) -> Result<(String, String)> {
        if passphrase.len() < 8 {
            return Err(BlockChainError::Crypto(
                "Passphrase must be at least 8 characters".into(),
            ));
        }

        let fname_private = format!("{filename_prefix}_private.pem");
        let fname_public = format!("{filename_prefix}_public.pem");

        let key_hex = hex::encode(self.key.to_bytes());
        let encrypted =
            CryptHelper::encrypt_with_passphrase(key_hex.as_bytes(), passphrase)?;
        encrypted.write_to_file(&fname_private, false)?;

        let pem = self.publickey().as_pem();
        std::fs::write(&fname_public, pem)?;

        Ok((fname_private, fname_public))
    }

    pub fn signing_key(&self) -> &SigningKey {
        &self.key
    }
}

// ---------------------------------------------------------------------------
// Hasher — SHA3-512 hasher with signing/verification
// ---------------------------------------------------------------------------

/// Wrapper around SHA3-512 that accumulates data and supports ECDSA signing.
///
/// Mirrors the Python Hasher class. Note: the Python `ecdsa` library internally
/// hashes the digest with SHA-256 before ECDSA signing (default for secp256k1).
/// k256 does the same when using the `Signer` trait, so behaviour matches.
pub struct Hasher {
    h: Sha3_512,
}

/// Things that can be fed into a Hasher.
pub enum Hashable<'a> {
    Bytes(&'a [u8]),
    Str(&'a str),
    HashRef(&'a Hash),
}

impl<'a> From<&'a [u8]> for Hashable<'a> {
    fn from(b: &'a [u8]) -> Self {
        Hashable::Bytes(b)
    }
}

impl<'a> From<&'a str> for Hashable<'a> {
    fn from(s: &'a str) -> Self {
        Hashable::Str(s)
    }
}

impl<'a> From<&'a Hash> for Hashable<'a> {
    fn from(h: &'a Hash) -> Self {
        Hashable::HashRef(h)
    }
}

impl<'a> From<&'a CryptoAddress> for Hashable<'a> {
    fn from(a: &'a CryptoAddress) -> Self {
        Hashable::Str(&a.address)
    }
}

impl Hasher {
    pub fn new() -> Self {
        Hasher { h: Sha3_512::new() }
    }

    pub fn new_with_message(msg: &[u8]) -> Self {
        let mut hasher = Self::new();
        hasher.h.update(msg);
        hasher
    }

    pub fn new_with_items(items: &[Hashable<'_>]) -> Self {
        let mut hasher = Self::new();
        for item in items {
            hasher.update_one(item);
        }
        hasher
    }

    pub fn update_one(&mut self, item: &Hashable<'_>) {
        match item {
            Hashable::Bytes(b) => self.h.update(b),
            Hashable::Str(s) => self.h.update(s.as_bytes()),
            Hashable::HashRef(h) => self.h.update(h.digest()),
        }
    }

    pub fn update_hash(&mut self, h: &Hash) {
        self.h.update(h.digest());
    }

    pub fn update_str(&mut self, s: &str) {
        self.h.update(s.as_bytes());
    }

    pub fn update_bytes(&mut self, b: &[u8]) {
        self.h.update(b);
    }

    pub fn get_hash(self) -> Hash {
        let result = self.h.finalize();
        Hash {
            data: result.to_vec(),
            bits: 512,
        }
    }

    /// Clone the internal state and produce a hash without consuming self.
    pub fn get_hash_clone(&self) -> Hash {
        let h = self.h.clone();
        let result = h.finalize();
        Hash {
            data: result.to_vec(),
            bits: 512,
        }
    }

    /// Sign the hash with the private key.
    ///
    /// The k256 `Signer` trait hashes the message with SHA-256 internally,
    /// matching the Python ecdsa library's default behaviour for secp256k1.
    pub fn sign(self, private_key: &ECDSAPrivateKey) -> Result<Hash> {
        let digest = self.h.finalize();
        let sig: Signature = k256::ecdsa::signature::Signer::sign(
            private_key.signing_key(),
            &digest,
        );
        let sig_bytes = sig.to_bytes();
        Ok(Hash {
            data: sig_bytes.to_vec(),
            bits: 512,
        })
    }

    /// Verify a signature against the public key.
    pub fn verify_signature(
        self,
        sig: &Hash,
        public_key: &ECDSAPublicKey,
    ) -> Result<()> {
        let digest = self.h.finalize();
        let signature = Signature::from_slice(sig.digest())
            .map_err(|e| BlockChainError::KeyValidation(format!("Invalid signature: {e}")))?;

        // k256 Verifier hashes the data with SHA-256 internally
        public_key
            .verifying_key()
            .verify(&digest, &signature)
            .map_err(|e| BlockChainError::KeyValidation(format!("Signature verification failed: {e}")))?;
        Ok(())
    }
}
